sigstore-java/src/main/java/dev/sigstore/tuf/Updater.java

```diff
@@ -225,82 +234,190 @@ void verifyDelegate(
+    if (expectedSha256 != null && !computedSha256.equals(expectedSha256)) {
+      badHashes.add(
+          new InvalidHashesException.InvalidHash("sha256", expectedSha256, computedSha256));
+    }
+    if (!badHashes.isEmpty()) {
+      throw new InvalidHashesException(
+          name, badHashes.toArray(InvalidHashesException.InvalidHash[]::new));
+    }
+  }
 
+  Targets updateTargets(Root root, Snapshot snapshot)
+      throws IOException, FileNotFoundException, InvalidHashesException,
+          SignatureVerificationException, NoSuchAlgorithmException, InvalidKeySpecException,
+          InvalidKeyException, FileExceedsMaxLengthException {
+    // 1) download the targets.json to max bytes
+    var targetsResultMaybe = fetcher.getMeta(Role.Name.TARGETS, Targets.class);
+    if (targetsResultMaybe.isEmpty()) {
+      throw new FileNotFoundException("targets.json", fetcher.getSource());
+    }
+    var targetsResult = targetsResultMaybe.get();
     // 2) check hash against snapshot.targets.hash, else fail.
-
+    verifyHashes(
+        "targets.json",
+        targetsResult.getRawBytes(),
+        snapshot.getSignedMeta().getTargetMeta("targets.json").getHashes());
     // 3) check against threshold of keys as specified by trusted root.json
-
+    verifyDelegate(root, targetsResult.getMetaResource());
     // 4) check targets.version == snapshot.targets.version, else fail.
-
+    int targetsVersion = targetsResult.getMetaResource().getSignedMeta().getVersion();
+    int snapshotTargetsVersion =
+        snapshot.getSignedMeta().getTargetMeta("targets.json").getVersion();
+    if (targetsVersion != snapshotTargetsVersion) {
+      throw new SnapshotVersionMismatchException(snapshotTargetsVersion, targetsVersion);
+    }
     // 5) check expiration is after tuf update start time
-
+    throwIfExpired(targetsResult.getMetaResource().getSignedMeta().getExpiresAsDate());
     // 6) persist targets metadata
+    // why do we persist the
+    localStore.storeMeta(targetsResult.getMetaResource());
+    return targetsResult.getMetaResource();
+  }
 
-    // 7) starting at each top level target role:
-    //        do pre-order DFS of metadata.
-    //        skip already visited roles
-    //        if maximum roles visited go to 8) downloading targets
-    //        process delegations (not sure if we need this yet)
-
-    // 8) If target is missing metadata fail.
-
-    // 9) Download target up to length specified in bytes. verify against hash.
-
-    // Done!!
-    return null;
+  void downloadTargets(Targets targets)
+      throws IOException, TargetMetadataMissingException, FileNotFoundException {
+    // Skip #7 and go straight to downloading targets. It looks like delegations were removed from
+    // sigstore TUF data.
+    // {@see https://github.com/sigstore/sigstore/issues/562}
+    // TODO(patrick@chainguard.dev): get review from Asra on whether we should be processing
```

Comment:
@asraa

Reply:
Yes, I'd skip delegations. Delegations have a tricky model (where higher targets files take precedence) and so we deferred these. We'll likely reintroduce this, but very slowly (starting with an optional one only relevant to npm, and they will probably set the starter example for their TUF client)

Reply:
ack.